src/engine/strat_engine/devlinks.rs

```
@@ -0,0 +1,118 @@
+
+use engine::Engine;
+
+pub const DEV_PATH: &str = "/dev/stratis";
+
+/// Set up directories and symlinks under /dev/stratis based on current
+/// config. Clear out other stuff.
+// Don't just remove and recreate everything in case there are processes
+// (e.g. user shells) with the current working directory within the tree.
+pub fn setup_devlinks(engine: &Engine) -> EngineResult<()> {
+    if let Err(err) = fs::create_dir(DEV_PATH) {
+        if err.kind() != ErrorKind::AlreadyExists {
+            return Err(From::from(err));
+        }
+    }
+
+    let mut existing_dirs: HashSet<_> = fs::read_dir(DEV_PATH)?
+        .map(|dir_e| {
+                 dir_e
+                     .expect("direntry is valid")
+                     .file_name()
+                     .into_string()
+                     .expect("Unix is utf8")
+             })
+        .collect();
+
+    for pool in engine.pools() {
+        if !existing_dirs.contains(pool.name()) {
+            pool_added(pool.name());
+        }
+        existing_dirs.remove(pool.name());
+
+        let pool_path: PathBuf = vec![DEV_PATH, pool.name()].iter().collect();
+
+        let mut existing_files: HashSet<_> = fs::read_dir(pool_path)?
+            .map(|dir_e| {
+                     dir_e
+                         .expect("direntry is valid")
+                         .file_name()
+                         .into_string()
+                         .expect("Unix is utf8")
+                 })
+            .collect();
+
+        for fs in pool.filesystems() {
+            if !existing_files.contains(fs.name()) {
+                filesystem_added(pool.name(), fs.name(), &fs.devnode());
+            }
+            existing_files.remove(fs.name());
+        }
+
+        for leftover in existing_files {
+            filesystem_removed(pool.name(), &leftover)
+        }
+    }
+
+    for leftover in existing_dirs {
+        pool_removed(str::from_utf8(leftover.as_bytes()).expect("is valid utf8"))
```

Better to return an error.